### PR TITLE
Agari data connector fixes and additions

### PR DESCRIPTION
**Pull request by udinagari**
- fixed typo in token gen if statement
- fixed header variable in APD call
- added x-header to API call to identify Sentinel
- added if statements for APD push to Sentinel to verify if there is data to push
- removed en-us from URLs in deployment template to pass document validation
- added BP logs to deployment template
-added log samples - raw is data from the API call, formatted is what is used to push to Sentinel
-added appropriate files to the zip archive

### CODE REVIEW

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 1]
```diff
@@ -182,20 +182,22 @@ if ($bpEnabled){
         $bpHeaders = @{
                     'Authorization' = "Bearer $bpToken"
                     'ContentType' = 'application/json'
+                    'x-api-requestby' = 'Sentinel/Powershell'
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Is this replacing the user-agent?
Is there a reason we cannot include a user-agent in the Agari API requests?

**Review by djspencer — COMMENTED** [review]
See inline comments / questions.
Also, do we need to include the .zip files in the repo?
Is this required by MS?

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 1 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Yes, they require it for the auto-deployment

Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Just re-tested. We can use 'UserAgent' = 'Sentinel/Powershell' or something like that - I was using the wrong string before

Comment (djspencer) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Would prefer to use a three-part string if possible. That is what we use in our other integrations.
The parts are:
app name
integration description / version
hosting application / version

E.g.
AgariSentinel  BP_APD_Integration/v1.0  SentinelPowerShell/v2.3

Ideally the Sentinel version could be acquired programatically.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 2]
```diff
@@ -83,6 +91,37 @@ Function Post-LogAnalyticsData($customerId, $sharedKey, $body, $logType)
...
+    if ($aprLogSuccess){
+        $newAppSettingList['APRlastLogTime'] = "t$nextStartDate"
+    }
+    if ($bpLogSuccess){
+        $newAppSettingList['BPlastLogTime'] = "t$nextStartDate"
+    }
+    # Write to App Settings if there are changes
+    if(($apdLogSuccess) -or ($aprLogSuccess) -or ($bpLogSuccess)){
+        Set-AzWebApp -ResourceGroupName $resGrp -Name $appName -AppSettings $newAppSettingList
+    }
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1:122]: We are using a single cursor for all three apps?
What happens if the API for one of the products does not complete?, or has and error?, or a 429 retry?

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 3]
```diff
@@ -182,7 +221,7 @@ if ($bpEnabled){
         $bpHeaders = @{
                     'Authorization' = "Bearer $bpToken"
                     'ContentType' = 'application/json'
-                    'x-api-requestby' = 'Sentinel/Powershell'
+                    'UserAgent' = 'AgariSentinel BP_APD_Integration/v1.0 SentinelPowerShellCore/v7'
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Can the integration to be running on different versions of SentinelPowerShellCore?
Better if we obtain the SentinelPowerShellCore version programmatically (of possible).

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 4]
```diff
@@ -197,7 +236,7 @@ if ($apdEnabled){
         $apdHeaders = @{
                     'Authorization'="Bearer $apdToken"
                     'ContentType' = 'application/json'
-                    'x-api-requestby' = 'Sentinel/Powershell'
+                    'UserAgent' = 'AgariSentinel BP_APD_Integration/v1.0 SentinelPowerShellCore/v7'
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Feels like we should include APR in the user agent if we are calling APR APIs.

A couple options here.
A single user-agent string link BP_APD_APR_Integration/v1.0.

Or a different one for each product API like:
BP_Integration/v1.0
APD_Integration/v1.0
APR_Integration/v1.0

I think I slightly prefer the later.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 5]
```diff
@@ -212,16 +251,18 @@ if ($aprEnabled){
...
     } else { 
         $aprToken = $false
         Write-Host "Error Fetching APR Token" 
         }
 }
+
 #Set global api variables
-$startdate=(Get-Date (get-date).addMinutes(-6) -UFormat "+%Y-%m-%dT%H:%M:01.00Z")
-$enddate=(Get-Date (get-date).addMinutes(-1) -UFormat "+%Y-%m-%dT%H:%M:00.00Z")
+$enddate=(Get-Date (get-date).addMinutes(-1) -UFormat "+%Y-%m-%dT%H:%M:00.000Z")
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1:266]: Set current time (get-date) once into a variable rather that calling get-time multiple times, which may cause slight drift.
e.g.: $oneMunuteAgo = (Get-Date (get-date).addMinutes(-1)

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 6]
```diff
@@ -232,12 +273,19 @@ $TimeStampField = "DateValue"
 
 #Check if BP is enabled and we got a token
 if (($bpEnabled) -and ($bpToken)) {
+    #Check to see if start time empty - usally first run, otherwise set the time from the env variable
+    if ($BPlastLog){
+        $startdate = $BPlastLog.substring(1)  
+    } else {
+        $startdate=(Get-Date (get-date).addMinutes(-6) -UFormat "+%Y-%m-%dT%H:%M:00:001Z")
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Why are we subtracting 6 minutes ?

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 7]
```diff
@@ -278,6 +325,12 @@ if (($bpEnabled) -and ($bpToken)) {
 
 #Check if APD is enabled 
 if (($apdEnabled) -and ($apdToken)){
+        #Check to see if start time empty - usally first run, otherwise set the time from the env variable
+        if ($apdlastLog){
+            $startdate = $APDlastLog.substring(1)  
+        } else {
+            $startdate=(Get-Date (get-date).addMinutes(-6) -UFormat "+%Y-%m-%dT%H:%M:00:001Z")
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Same Q as above. Plus again, best to use a singe time reference stored in a variable.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 8]
```diff
@@ -322,6 +375,12 @@ if (($apdEnabled) -and ($apdToken)){
 if ($sgEnabled){
     # If APR enabled, get all malicious IoC verdicts from investigations 
     if (($aprEnabled) -and ($aprToken)){
+        #Check to see if start time empty - usally first run, otherwise set the time from the env variable
+        if ($aprlastLog){
+            $startdate = $APRlastLog.substring(1)  
+        } else {
+            $startdate=(Get-Date (get-date).addMinutes(-6) -UFormat "+%Y-%m-%dT%H:%M:00:001Z")
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1]: Same a above.

**Review by djspencer — CHANGES_REQUESTED** [review]
I have questions and suggestions around your cursor management.
Also, what kind of test framework is there for Sentinel?
There really should have tests for all this code! I.e. follow TDD pattern.
I know we are on a time timeline to get this out, but next time we need to include tests in the development plan. Really helps flush out use case and corner case conditions, which is very common with cursor management.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 3 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: There isn't a variable that can be easily scripted, powershell 7 is what runs in azure.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 4 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: OK, I'll change that in the next commit.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 5 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1:266]: $enddate is only created once. I see the issue with the startdate so I'll take them out of the individual products calls.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 6 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: To get a full 5 minutes (i.e startdate is 6 mins in the past and the enddate is 1 min in the past. I did this to account for any potential delays in the timer script.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 7 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: ack.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 8 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1]: ack.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 (hunk 2 above)
Comment (udinagari) [reply DataConnectors/Agari/AzureFunctionAgari/run.ps1:122]: If we get a 4xx error, the invoke-restapi won't create any psobjects so the foreach loop won't trigger thus the logsuccess won't flip to true. The next run will fetch from whatever time is in the environment variable, so it won't have any gaps in the logs. if it does run, we want to populate the env variable with the next start date. I have this running in sentinel now, with our agari data, and it's been running successfully for the last week.  We need to get this to MS so that we can get the private preview where we can watch / assess with 2 customers.  In the future we should consider putting a "api last accessed field in the APIs so we don't have to rely on the app logging fails / success data requests.

File: DataConnectors/Agari/AzureFunctionAgari/run.ps1 [hunk 9]
```diff
@@ -221,7 +221,7 @@ if ($bpEnabled){
         $bpHeaders = @{
                     'Authorization' = "Bearer $bpToken"
                     'ContentType' = 'application/json'
-                    'UserAgent' = 'AgariSentinel BP_APD_Integration/v1.0 SentinelPowerShellCore/v7'
+                    'UserAgent' = "AgariSentinel BP_Integration/v1.0 SentinelPowerShellCore/v$PSVersionTable.PSVersion.major"
```
Comment (djspencer) [inline DataConnectors/Agari/AzureFunctionAgari/run.ps1:224]: Thanks for getting the version programmatically?
Curious - Does Sentinel have a version?
(no need to change this at this point)

**Review by djspencer — APPROVED** [review]
